src/general_tally.cpp

```diff
@@ -0,0 +1,243 @@
+#include <tallies/general_tally.hpp>
+#include <utils/error.hpp>
+#include <utils/output.hpp>
+
+#include <boost/container/static_vector.hpp>
+using StaticVector4 = boost::container::static_vector<size_t, 4>;
+
+GeneralTally::GeneralTally(std::shared_ptr<PositionFilter> position_filter,
+                           std::shared_ptr<EnergyFilter> energy_in,
+                           Quantity quantity, Estimator estimator,
+                           std::string name_)
+    : ITally(quantity, estimator, name_),
+      position_filter_(position_filter),
+      energy_in_(energy_in) {
+  // at least one filter must exists.
+  if (position_filter_ == nullptr && energy_in_ == nullptr) {
+    fatal_error("for the " + name_ + " tally, no filter is provided.");
+  }
+
+  // get the shape or dimension of position filter
+  StaticVector3 position_shape_;
+  if (position_filter_) {
+    position_shape_ = position_filter_->get_shape();
+  }
+
+  StaticVector4 tally_dimensions_(position_shape_.begin(),
+                                  position_shape_.end());
+
+  // get the size of the enery_filter
+  if (energy_in_) {
+    size_t ne = energy_in_->size();
+    tally_dimensions_.insert(tally_dimensions_.begin(), ne);
+  }
```

Comment:
We should add the energy portion to `tally_dimensions_` first, so that we don't need to insert it at the beginning, copying all the others elements from the `position_shape_`.